dbms/src/Functions/FunctionsLogical.h

```diff
@@ -377,154 +504,421 @@ class FunctionAnyArityLogical : public IFunction
+                nullable_uint8_columns.push_back(col_res.get());
+                null_maps.push_back(&vec_res_is_null);
+            }
+            if (nullable_uint8_columns.size() == 1)
+            {
+                // special case
+                assert(!has_not_null_column);
+                vec_res.assign(nullable_uint8_columns[0]->getData());
+                vec_res_is_null.assign(*null_maps[0]);
+            }
             else
             {
-                auto converted_column = ColumnUInt8::create(rows);
-                convertToUInt8(column, converted_column->getData(), vec_res_not_null, vec_input_has_null);
-                uint8_in.push_back(converted_column.get());
-                converted_columns.emplace_back(std::move(converted_column));
+                while (nullable_uint8_columns.size() > 1)
+                {
+                    NullableAssociativeOperationImpl<Impl, 6>::execute(
+                        nullable_uint8_columns,
+                        null_maps,
+                        vec_res,
+                        vec_res_is_null);
+                    nullable_uint8_columns.push_back(col_res.get());
+                    null_maps.push_back(&vec_res_is_null);
+                }
             }
         }
 
-        /// Effeciently combine all the columns of the correct type.
-        while (uint8_in.size() > 1)
+        if (has_const)
         {
-            /// With a large block size, combining 6 columns per pass is the fastest.
-            /// When small - more, is faster.
-            AssociativeOperationImpl<Impl, 10>::execute(uint8_in, vec_res);
-            uint8_in.push_back(col_res.get());
+            // if there is const, then apply the constant at last
+            if (result_is_nullable)
+            {
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    std::tie(vec_res[i], vec_res_is_null[i])
+                        = Impl::applyTwoNullable(const_val, const_val_is_null, vec_res[i], vec_res_is_null[i]);
+                }
+            }
+            else
+            {
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    vec_res[i] = Impl::apply(const_val, vec_res[i]);
+                }
+            }
         }
 
-        /// This is possible if there is exactly one non-constant among the arguments, and it is of type UInt8.
-        if (uint8_in[0] != col_res.get())
-            vec_res.assign(uint8_in[0]->getData());
-
-        if constexpr (!special_impl_for_nulls)
-            block.getByPosition(result).column = std::move(col_res);
+        if (result_is_nullable)
+        {
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_res), std::move(col_res_is_null));
+        }
         else
         {
-            if (has_nullable_input_column)
+            block.getByPosition(result).column = std::move(col_res);
+        }
+    }
+
+    void convertAllInputToUInt8(
```

Comment:
Better add comments to decribe the relations among output columnPtrs(not_null_uint8_columns, nullable_uint8_columns, converted_columns)

Reply:
Done